### ISSUE TITLE
VersionBits (BIP 9) and AuxPoW

### ISSUE THREAD

**Issue by JeremyRand** [dogecoin/dogecoin#1334]
Hello!

I figure I should reach out to you about some changes that Namecoin is looking at, to see whether this will affect you, whether it's on your radar, and whether maybe you already have a solution in place.

Bitcoin's BIP 9 (VersionBits) conflicts with the way Namecoin's AuxPoW works.  Specifically, VersionBits permits values of nVersion in block headers which may make those blocks invalid as merge-mined blocks, or as parent blocks of merge-mined blocks, because AuxPoW implementations recognize those values as having unexpected values of the AuxPoW bit or the AuxPoW Chain ID.

Namecoin is planning to do a hardfork in the foreseeable future to deal with this.  You can read up on Namecoin's plans in this forum thread: https://forum.namecoin.org/viewtopic.php?f=5&t=2466 .  I'm happy to provide additional details as needed.

I have no idea how urgent this is for you, since you're not using Bitcoin as your AuxPoW parent chain, but I figure it's a good idea to make sure you're aware of the issue, and to get your input on the subject.

Cheers,

-Jeremy Rand, namecoin.org

**Comment by patricklodder** [comment]
Thank you for the ping, appreciated. There's always a good chance that Litecoin will port such features 1-on-1 from Bitcoin, so we might run into the same issues (if Bitcoin plans on introducing this with a softfork, we might see the problem even sooner in a live situation, as Litecoin softforks are rather quick processes.)

**Comment by patricklodder** [comment]
Just to summarize the discussion over at the namecoin forum:

The idea is to perform 2 changes:
1. only accept AuxPoW-enabled headers
2. move chain identification from the version property to the nonce property. 

Direct Impact:
- Non-merged mining is still possible, but must be performed using a (dummy) auxpow structure, like we do at the [auxpow rpc test suite](https://github.com/dogecoin/dogecoin/blob/master/qa/rpc-tests/test_framework/auxpow.py#L12). At block 1109964 on mainnet, I have not encountered a non-auxpow block for the last 50k blocks, so this would only impact blockchain size on testnet significantly. This means we'll have to rewrite features like `src/miner.cpp` and it's unit tests.
- The [`Aux POW parent cannot have our chain ID`](https://github.com/dogecoin/dogecoin/blob/master/src/auxpow.cpp#L113-L114) constraint will become obsolete with this because dogecoin PoW will never match on the scrypt hash in the blockheader.
- We can do this with a softfork.

Indirect impact:
- Since we didn't softfork for CLTV yet, we might need to do those at the same time. We could  solve this by making CLTV a softfork on top of this one, with a release-only config option for miners to mine at version 4 (auxpow softfork without CLTV) or version 5 (auxpow + CLTV softfork)? 

@langerhans @rnicoll Thoughts?

**Comment by langerhans** [comment]
So after a long explaining session I agree with the proposal. Pretty sure v5 should be default in this case though since we should aim to get those forks all through. Ideally they activate at the same time and I''m pretty sure they will cause CLTV is not _that_ controversial.

**Comment by rnicoll** [comment]
Only thing I'm wondering about is why we don't put chain ID into the coinbase TX rather than trying to put it into the header.

**Comment by JeremyRand** [comment]
@patricklodder So I'm curious why this is a softfork for you.  As far as I can tell, the BIP 9 changes allow values of the merged-mined chain's block version which are disallowed by the current AuxPoW rules, making it a hardfork.  (In particular, the current rules reject a value of the top 16 bits that don't match the chain ID, whereas such values are allowed under BIP 9.)  Am I missing something?

@rnicoll I'm actually not certain why Namecoin is planning to put the chain ID in the header rather than the coinbase.  Namecoin's lead C++ developer @domob1812 might have thoughts on that topic.

Cheers.  (Oh, and thanks for the Twitter shout-out a few days ago.  Always good to remind end users that we're friends.  :) )

**Comment by domob1812** [comment]
Yes, I would also be interested to know how that would just be a softfork.  BTW, I've already implemented the "mine all blocks using auxpow" thing mentioned above (with a dummy auxpow structure), see https://github.com/namecoin/namecoin-core/pull/66.

In principle you don't have to put the chain ID (or any other information for that matter) into the block at all, since the requirement that all blocks must be merge mined already prevents a miner from using a found PoW both for a block and for merge-mining a second one.  (Unless I miss something, that is, so feel free to comment if you think otherwise.)

However, in general it makes sense to have the chain ID available; among other things, this allows coins with multiple PoW algos, and cannot hurt in general.  Having the information in the header allows it to be used for header-first validation or SPV, where you don't have the coinbase available.  The `nNonce` field is free for that purpose and is unlikely to clash with future Bitcoin changes since Bitcoin needs it for the nonce.

**Comment by patricklodder** [comment]
@JeremyRand You're right about the constraint removal having to be a hardfork, we can indeed not softfork for that. Thanks for thinking with us because I would've likely ran into this while coding, or worse, writing the qa code.

So, since we learned at the last release that it takes about 5 months to get ~51% of nodes (not miners) updated (at least that's what my nodes stats say, perhaps someone with a dns seed has more detailed info on this) the best we can do is do a conditional hardfork over a long (6 months or so) time period? So we softfork the auxpow-only and nonce=chainId changes (those can softfork as long as they are relative to a block version) and only if that goes into super-majority, we hardfork out the constraint and auxpow version bitmasks at block/time N? I really don't want us (read: rnicoll) to do another round of "let's call everybody" like we did at the introduction of AuxPoW, also because we had nearly 10% of our pools running in fire-and-forget mode at our last softfork, so it makes more sense to me to filter those out with a soft super-majority mechanism first.

With the hardfork in mind, we'll have to act quickly: if bip9 will be included in Bitcoin 0.13 then we can count on it going into Litecoin shortly after that, so 6 months is already tight...
